Release notes draft for V0.6.0 (Tungsten Release) (#166)
dfinity/docs: modules/release-notes/pages/sdk-release-notes.adoc

The {release} release includes the following new features and capabilities:

=== SDK

- You can now look up a canister identifier using the command `dfx canister id <canister_name>`.
- The `--check` flag can be used with the `dfx build` command to check whether a canister will build before creating or building the canister.
- Both canister name and identifiers are printed upon canister creation.
- Canister manifest now stores multiple canister IDs associated with local and remote networks as well as network mapping.

> - Asset canisters include `.did.js`.

Reviewer: This is too internal.

- Local network defaults to `127.0.0.1:8000`.

=== Tungsten (Developer Network)

- HTTP authorization and credentials management for onboarded Tungsten users.
- `Tungsten` added as a provider in `dfx.json`
- You can use the `dfx ping` command to access and toggle between local and Tungsten networks.
- The `--network <network>` flag can be used to build and install canisters to the specified provider.
- Canister ID formatting for accessing Tungsten-deployed apps in the browser.

=== Motoko

- The `motoko-base` repository is now open. We encourage developers to use Vessel package manager to download the latest `base` from `master`.

> - Major breaking changes and updates as detailed here: dfinity/motoko-base#37

Reviewer: ```suggestion
- Major breaking changes and updates as detailed here: https://github.com/dfinity/motoko-base/issues/37
```
Do we want to move this to the breaking change section?

- Motoko style guide published to sdk.dfinity.org
- Stable variable support

> 

Reviewer: ```suggestion
- Released `mo-doc` for generating documentation from Motoko comments
- Better support for `Char` and `Text` module
- `Error` module for rejecting messages
- `Buf` module renamed to `Buffer`
```

== Breaking changes
The {release} release includes the following changes that might require updates to existing programs:

- The command `dfx new` now creates a separate assets canister. Programs built with earlier versions of the SDK may need to be converted to this new format. 
- You must now create empty canisters before building and installing using the subcommand `dfx canister create`.
- `dfx canister call` will consult the candid file for method types. There is no need to use `--type string/number` anymore. The arguments are therefore deprecated.

== Issues fixed in this release

This section covers the issues fixed in this release.
The {release} release includes internal fixes and improvements to the Candid user interface, the interface description library, and some refactoring of `dfx` commands.

- `dfx stop` now finds and kills all `dfx start` and `dfx replica` processes.
- Allow lowercase hex in Canister ID.

> - Canister name now used for WASM and DID output.

Reviewer: This is internal

- Allow installation without sudo when possible.
- Install script issues resolved for Ubuntu and Mac.
- Check added to forbid starting webserver with a forwarded port.
- Cache directory management and error messages.

== Known issues and limitations

This section covers any known issues or limitations that might affect how you work with the {sdk-short-name} in specific environments or scenarios.
If there are workarounds to any of the issues described in this section, you can find them in the link:../developers-guide/troubleshooting{outfilesuffix}[Troubleshooting] section.

> The {release} release includes the following known issues and limitations:
> 

Reviewer: ```suggestion
```

The {release} release includes the following new features and capabilities:

=== SDK

- You can now look up a canister identifier using the command `dfx canister id <canister_name>`
- The `--check` flag can be used with the `dfx build` command to check whether a canister will build before creating or building the canister
- Both canister name and identifiers are printed upon canister creation
- Canister manifest now stores multiple canister IDs associated with local and remote networks as well as network mapping
- Local network defaults to `127.0.0.1:8000`

=== Tungsten (Developer Network)

- HTTP authorization and credentials management for onboarded Tungsten users
- `Tungsten` added as a provider in `dfx.json

> - You can use the `dfx ping` command to access and toggle between local and Tungsten networks

Reviewer 2: `dfx ping` doesn't toggle between local and Tungsten. I think we should just use the help string here:
`Ping an Internet Computer and returns its status`

Author: Fixed. @p-shahi should we note this as a known issue? That you can't currently toggle?

Reviewer 2: don't think it's actually a feature to toggle with `dfx ping` so we don't need to note it @alexabsmith

- Canister ID formatting for accessing Tungsten-deployed apps in the browser.

=== Motoko

- Stable variable support
- Improved type checking of literals as payload for variant tags. The following will now type-check:
`func () : { #tag : Nat8 } = #tag 42;`
- Warning when the read file has a different name from the requested file.
- `debug_show` implemented for `Principal`

> - Type hash implementation

Reviewer: I would say these are all very small/internal changes that can be omitted. We are missing the really BIG BREAKING changes: https://github.com/dfinity/motoko-base/issues/37

Reviewer: It's also worth mentioning the open source of `base` repo, and encourage developers to use vessel to download the latest base from master.

Reviewer: @lsgunnlsgunn Will we publish the style guide (https://github.com/dfinity-lab/motoko/blob/master/doc/style.md) on the SDK website along with the dfx release? If so, it would be great to promote the guide here.

Author: @chenyan-dfinity @lsgunnlsgunn @stanleygjones would it be appropriate to direct folks to that PR now that the repo is open for a summary of the major changes? Vs. listing them out here?

Reviewer: Sounds good to me. It's generally good to promote the base repo, and encourage people to check there more often. Note that there are other changes other than this issue. The git logs in the repo is really messy, so it's hard to dig them out....

== Breaking changes
The {release} release includes the following changes that might require updates to existing programs:

- The command `dfx new` now creates a separate assets canister. Programs built with earlier versions of the SDK may need to be converted to this new format. 
- You must now create empty canisters before building and installing using the subcommand `dfx canister create`.

> - `string` and `number` are no longer supported as arguments to the `--type` flag for `dfx canister call` commands.

Reviewer: ```suggestion
- `dfx canister call` will consult the candid file for method types. There is no need to use `--type string/number` anymore. The arguments are therefore deprecated.
```

> - Methods starting with `__` will break, as those are not implemented anymore.

Reviewer: Where does this come from? I'm not aware of this change.

Author: @chenyan-dfinity I dug it out of this PR, any ideas? https://github.com/dfinity-lab/sdk/pull/804

Reviewer: I see. These are very internal changes. It doesn't affect external developers.

Author: Removed.

> - When creating an actor, you need to pass in the canister ID.

Reviewer: Also not true?

-- Network mapping for `local` and `Tungsten`
-- Local network defaults to `127.0.0.1:8000`.

=== Tungsten (Developer Network)

- HTTP authorization and credentials management for onboarded Tungsten users
- `Tungsten` added as a provider in `dfx.json
- You can use the `dfx ping` command to access and toggle between local and Tungsten networks
- The `--network <network>` flag can be used to build and install canisters to the specified provider
- Canister ID formatting for accessing Tungsten-deployed apps in the browser

=== Motoko

- The `motoko-base` repository is now open. We encourage developers to use Vessel package manager to download the latest `base` from `master`.
- Motoko style guide published to sdk.dfinity.org
- Stable variable support
- Released `mo-doc` for generating documentation from Motoko comments
- Better support for `Char` and `Text` module
- `Error` module for rejecting messages
- `Buf` module renamed to `Buffer`

== Breaking changes
The {release} release includes the following changes that might require updates to existing programs:

> - Major breaking changes and updates to Motoko as detailed here: dfinity/motoko-base#37

Reviewer: ```suggestion
- Major breaking changes and updates to Motoko as detailed here: https://github.com/dfinity/motoko-base/issues/37
```
